experimental/alp/lib/Transforms/extract_kernel_pass.cpp

```diff
@@ -192,147 +225,140 @@ std::unique_ptr<mlir::OperationPass<ModuleOp>> mlir::createExtractKernelPass() {
+      return false;
+    });
+  }
+
   // Add a returnOp into the block to properly terminate it
   rewriter.setInsertionPointToEnd(entry_block);
-  //rewriter.create<ReturnOp>(loc);
+  // rewriter.create<ReturnOp>(loc);
 
   // We are done with the callee. Now we have to work on the caller. The overall
   // idea is to insert a new_block right before the successor of the old block.
   // If the old block has no successors, then add it at the end of the region
 
   rewriter.setInsertionPointToEnd(origin);
 
-  // Create the call 
+  // Create the call
   rewriter.create<CallOp>(loc, func_op, vals);
   rewriter.create<ReturnOp>(loc);
 }
 
-// Walk the for loops and find the one that as operands. In GEMM is the micro-kernel.
-// TODO: we should have the linalg::split to signal the microkernel of the operation
-// and use it to run the function extractor if needed
-struct ExtractKernelTailPass : public ExtractKernelTailPassBase<ExtractKernelTailPass> {
+// Walk the for loops and find the one that as operands. In GEMM is the
+// micro-kernel.
+// TODO: we should have the linalg::split to signal the microkernel of the
+// operation and use it to run the function extractor if needed
+struct ExtractKernelTailPass
+    : public ExtractKernelTailPassBase<ExtractKernelTailPass> {
 
   ExtractKernelTailPass() = default;
-  ExtractKernelTailPass(const ExtractKernelTailPass& pass) { }
-  void getDependentDialects(DialectRegistry &registry) const override {
-
-  }
+  ExtractKernelTailPass(const ExtractKernelTailPass &pass) {}
+  void getDependentDialects(DialectRegistry &registry) const override {}
   void runOnOperation() override {
     // Get the current FuncOp operation being operated on.
     auto module = getOperation();
     scf::ForOp loop;
+    LLVM_DEBUG(llvm::dbgs() << "extract_kernel_tail starts\n ");
 
     for (FuncOp func : module.getOps<FuncOp>()) {
       // Walk the operations within the function.
       func.walk([&](scf::ForOp forop) {
-        if (forop.getNumIterOperands()){
+        if (forop.getNumIterOperands()) {
           loop = forop;
         }
       });
     }
 
     IRRewriter rewriter(module.getContext());
-    Block *tail = rewriter.splitBlock(loop->getBlock(), Block::iterator(loop->getNextNode()));
-    //rewriter.setInsertionPointToEnd(loop->getBlock());
-    // Create the call 
-    //rewriter.create<CallOp>(loc, func_op, vals);
-    //rewriter.create<ReturnOp>(module.getLoc());
-    extract_function_2("kernel_tail", tail, module, rewriter, loop->getBlock(), module.getLoc());
-    
+    Block *tail = rewriter.splitBlock(loop->getBlock(),
+                                      Block::iterator(loop->getNextNode()));
+    extract_function_2("kernel_tail", tail, module, rewriter, loop->getBlock(),
+                       module.getLoc());
   }
 };
 
-std::unique_ptr<mlir::OperationPass<ModuleOp>> mlir::createExtractKernelTailPass() {
+std::unique_ptr<mlir::OperationPass<ModuleOp>>
+mlir::createExtractKernelTailPass() {
```

Comment:
I would be interested in synchronizing on what this does and whether it can be subsumed by peeling + the recent transform to outline some parent loop given an op.
My hope is we could replace this file with more standard stuff but I am not sure I have the details right.

Reply:
We don't use this pass anymore. Basically, using a different scheduling algorithm, this pass can be (fortunately discarded). Once I'll do a bit of c++ refactoring + testing I will remove it

Reply:
Great!